File: src/main/java/org/algo4j/tree/heap/Heap.java
package org.algo4j.tree.heap;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

/**
 * Created by ice1000 on 2017/1/31.
 *
 * @author ice1000
 */
public interface Heap extends Cloneable {
	@Contract(pure = true)
	int getElementAt(int index);

	@Contract(pure = true)
	int size();

	@Contract(pure = true)
	boolean empty();

	void push(int element);

	default int peek() {
		int ret = top();
		pop();
		return ret;
	}

	@Contract(pure = true)
	int top();

	void pop();

	@NotNull
	@Contract(pure = true)
	MinHeap clone();
}

File: src/main/kotlin/org/algo4j/Heap.kt
@file:JvmName("Extension")
@file:JvmMultifileClass

/**
 * Created by ice1000 on 2017/1/24.
 *
 * @author ice1000
 */
package org.algo4j

import org.algo4j.tree.heap.MaxHeap
import org.algo4j.tree.heap.MinHeap
import org.jetbrains.annotations.Contract


@Contract(pure = true)
operator fun MinHeap.get(index: Int): Int =
		getElementAt(index)

@Contract(pure = false)
operator fun MinHeap.plusAssign(value: Int) =
		push(value)

/**
 * functional plus
 */
@Contract(pure = true)
operator fun MinHeap.plus(value: Int) = clone().apply { plusAssign(value) }


@Contract(pure = true)
operator fun MaxHeap.get(index: Int): Int =
		getElementAt(index)

@Contract(pure = false)
operator fun MaxHeap.plusAssign(value: Int) =
		push(value)

/**
 * functional plus
 */
@Contract(pure = true)
operator fun MaxHeap.plus(value: Int) = clone().apply { plusAssign(value) }

